### PR TITLE
Oak 11246 - Introduce config for perflogger to log very slow operations at info

### PR DESCRIPTION
**Pull request by ionutzpi**
Introduce config for perflogger to log very slow operations at info

### CODE REVIEW

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 1]
```diff
@@ -320,6 +321,13 @@
                     " (milliseconds).")
     long recoveryDelayMillis() default DEFAULT_RECOVERY_DELAY_MILLIS;
 
+    @AttributeDefinition(
+            name = "Perflogger Info Millis",
+            description = "Minimum duration (in milliseconds) for an operation that perflogger info will log. " +
+                    " Default: " + DEFAULT_PERFLOGGER_INFO_MILLIS +
+                    " (milliseconds).")
+    long perfloggerInfoMillis() default DEFAULT_PERFLOGGER_INFO_MILLIS;
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]:
```suggestion
    long perfLoggerInfoMillis() default DEFAULT_PERFLOGGER_INFO_MILLIS;
```

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 1 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 2]
```diff
@@ -320,6 +321,13 @@
                     " (milliseconds).")
     long recoveryDelayMillis() default DEFAULT_RECOVERY_DELAY_MILLIS;
 
+    @AttributeDefinition(
+            name = "Perflogger Info Millis",
+            description = "Minimum duration (in milliseconds) for an operation that perflogger info will log. " +
+                    " Default: " + DEFAULT_PERFLOGGER_INFO_MILLIS +
+                    " (milliseconds).")
+    long perfLoggerInfoMillis() default DEFAULT_PERFLOGGER_INFO_MILLIS;
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:335]: could you please add a unit case for this?

**Comment by rishabhdaim** [comment]
could you please add unit cases for classes that had been modified?

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 2 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:335]: Done

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/PerfLogger.java [hunk 3]
```diff
@@ -50,6 +50,8 @@ public final class PerfLogger {
     /** The logger to which the log statements are emitted **/
     private final Logger delegate;
 
+    private long infoLogMillis;
```
Comment (stefan-egli) [inline oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/PerfLogger.java]: this should not be zero by default. with a zero as default (and without ensuring that all PerfLoggers used throughout the code base call `setInfoLogMillis` with their desired value) it would log everything - which would cause a huge amount of logs. I would suggest the same value as was already used prior to this PR, namely Long.MAX_VALUE

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/AbstractDocumentNodeState.java [hunk 4]
```diff
@@ -162,4 +162,8 @@ private boolean revisionEquals(AbstractDocumentNodeState other) {
         return this.getLastRevision() != null
                 && this.getLastRevision().equals(other.getLastRevision());
     }
+
+    static void configurePerfLogger(long infoLogMillis) {
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/AbstractDocumentNodeState.java:171]: I would prefer a short javadoc to explain what this does

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/ClusterNodeInfo.java [hunk 5]
```diff
@@ -240,6 +240,9 @@ static RecoverLockState fromString(String state) {
     /** OAK-10281 : default millis to delay a recovery after a lease timeout */
     static final long DEFAULT_RECOVERY_DELAY_MILLIS = 0;
 
+    /** OAK-11246 : default millis for perflogger info */
+    static final long DEFAULT_PERFLOGGER_INFO_MILLIS = Long.MAX_VALUE;
+
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/ClusterNodeInfo.java]: don't think this belongs here, as it has nothing to do with ClusterNodeInfo. I would suggest to move it to a more generic place, eg to DocumentNodeStoreService

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 6]
```diff
@@ -320,6 +321,13 @@
                     " (milliseconds).")
     long recoveryDelayMillis() default DEFAULT_RECOVERY_DELAY_MILLIS;
 
+    @AttributeDefinition(
+            name = "Perflogger Info Millis",
+            description = "Minimum duration (in milliseconds) for an operation that perflogger info will log. " +
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]:
```suggestion
            description = "Minimum duration (in milliseconds) for certain operations that perflogger info will log. " +
```
perhaps we can emphasize that it is only "certain" operations (while not making a definition of what certain exactly means, that can also change over time) are perf-logged. otherwise it might imply that perflogger in general would use this (which might or might not be the case)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java [hunk 7]
```diff
@@ -152,7 +152,7 @@ public final class DocumentNodeStore
 
     private static final Logger LOG = LoggerFactory.getLogger(DocumentNodeStore.class);
 
-    private static final PerfLogger PERFLOG = new PerfLogger(
+    static final PerfLogger PERFLOG = new PerfLogger(
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: Quite liked the approach of the the other two cases where a new static `configurePerfLogger` is introduced. Avoids requiring this static to be made open.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java [hunk 8]
```diff
@@ -860,4 +860,8 @@ NodeState merge(@NotNull CommitHook hook,
             throw ex;
         }
     }
+
+    static void configurePerfLogger(long infoLogMillis) {
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java:869]: (same as above: would suggest a small javadoc here)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java [hunk 9]
```diff
@@ -377,6 +379,10 @@ private void registerNodeStore() throws IOException {
                 newArrayList(gcMonitor, loggingGCMonitor)));
         mkBuilder.setRevisionGCMaxAge(TimeUnit.SECONDS.toMillis(config.versionGcMaxAgeInSecs()));
 
+        PERFLOG.setInfoLogMillis(config.perfLoggerInfoMillis());
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: (see above: consider also using a `configurePerfLogger` here)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java [hunk 10]
```diff
@@ -343,6 +344,7 @@ private void registerNodeStore() throws IOException {
             builder.setLeaseSocketTimeout(config.mongoLeaseSocketTimeout());
             builder.setMongoDB(uri, db, config.blobCacheSize());
             builder.setCollectionCompressionType(config.collectionCompressionType());
+            builder.setPerfloggerInfoMillis(config.perfLoggerInfoMillis());
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: either this one or the one in `configureBuilder` is a duplicate. (perhaps this one is the duplicate as the other one is also used in a few similar static cases already?)

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java [hunk 11]
```diff
@@ -239,6 +239,17 @@ public void overrideFullGCExcludePaths() throws Exception {
         assertArrayEquals(new String[]{"/foo", "/bar"}, config.fullGCExcludePaths());
     }
 
+    @Test
+    public void perfLoggerInfoMillisTest() throws Exception {
+        addConfigurationEntry(preset, "perfLoggerInfoMillis", 0L);
```
Comment (stefan-egli) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java]: I would suggest to add a test for the default (which as mentioned above should not be zero in my view)

**Review by stefan-egli — COMMENTED** [review]
a few comments added:

File: oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/PerfLogger.java (hunk 3 above)
Comment (ionutzpi) [reply oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/PerfLogger.java]: I set it Long.MAX_VALUE. Thanks

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/AbstractDocumentNodeState.java (hunk 4 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/AbstractDocumentNodeState.java:171]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/ClusterNodeInfo.java (hunk 5 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/ClusterNodeInfo.java]: Done. Moved var to DocumentNodeStoreService.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 6 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: Done.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java (hunk 7 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java (hunk 8 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBranch.java:869]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 9 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: Done

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 10 above)
Comment (ionutzpi) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java]: Removed this one. I kept it in configureBuilder.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java (hunk 11 above)
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java]: Done.

**Comment by rishabhdaim** [comment]
@ionutzpi could you please resolve the conflicts?

**Comment by ionutzpi** [comment]
Done.

**Review by stefan-egli — APPROVED** [review]
+1, lgtm.

Not sure why the test fail - would like to test locally first to see if it's unrelated - from a quick look it seems unrelated.. but might be good to double-check before merging

**Review by rishabhdaim — APPROVED** [review]

**Comment by stefan-egli** [comment]
@rishabhdaim , I think we should generally squash merge PRs - otherwise they unnecessarily fill our trunk commit history with a large amount of smaller changes that are not really needed. Wdyt?

**Comment by rishabhdaim** [comment]
@stefan-egli I agree that we could squash and merge small PRs (like the above).
